src/parsi-c.cpp

```diff
@@ -0,0 +1,382 @@
+#include "parsi/parsi-c.h"
+
+#include <cstddef>
+#include <cstdlib>
+#include <cstring>
+
+struct parsi_compiled_parser {
+    parsi_parser_t* parser;
+};
+
+static bool parsi_charset_check(parsi_charset_t* charset, char chr)
+{
+    const std::size_t idx = static_cast<std::size_t>(chr);
+    const std::size_t bit = static_cast<std::size_t>(1) << (idx & 63ull);  // idx % 64
+    return charset->bitset[idx >> 6ull] & bit;  // idx / 64
+}
+
+parsi_charset_t parsi_charset(const char* str)
+{
+    return parsi_charset_n(str, std::strlen(str));
+}
+
+parsi_charset_t parsi_charset_n(const char* str, size_t size)
+{
+    std::size_t bytes[4] = {0};
+    for (std::size_t index = 0; index < size; ++index) {
+        const unsigned char chr = static_cast<unsigned char>(str[index]);
+        const std::size_t bit = static_cast<std::size_t>(1) << (chr & 63);  // chr % 64
+        bytes[chr >> 6] |= bit;  // chr / 64
+    }
+    return parsi_charset_t{ .bitset = { bytes[0], bytes[1], bytes[2], bytes[3] } };
+}
+
+parsi_parser_t* parsi_alloc_parser(parsi_parser_t parser)
```

Comment:
The `std::malloc` call could fail, returning `nullptr`. It's good practice to check if `malloc `was successful before using the pointer.